File: Merge sorting algorithm/MergeSort.java
public class MergeSort {

    public static int[] merge(int[] arr1, int[] arr2) {

        int arr1_idx = 0;
        int arr2_idx = 0;

        int[] sorted_arr = new int[arr1.length + arr2.length];

        int idx = 0;
        while (arr1_idx < arr1.length && arr2_idx < arr2.length) {
            if (arr1[arr1_idx] <= arr2[arr2_idx]) {
                sorted_arr[idx] = arr1[arr1_idx];
                arr1_idx += 1;
            } else {
                sorted_arr[idx] = arr2[arr2_idx];
                arr2_idx += 1;
            }
            idx += 1;
        }

        while (arr1_idx < arr1.length) {
            sorted_arr[idx] = arr1[arr1_idx];
            arr1_idx += 1;
            idx += 1;
        }

        while (arr2_idx < arr2.length) {
            sorted_arr[idx] = arr2[arr1_idx];
            arr2_idx += 1;
            idx += 1;
        }

        return sorted_arr;
    }

    public static int[] sort(int[] arr) {
        int n = arr.length;
        if (n <= 1) {
            return arr;
        }

        int mid = n / 2;
        int[] left_arr = new int[mid];
        int[] right_arr = new int[n - mid];

        for (int i = 0; i < mid; i++) {
            left_arr[i] = arr[i];
        }

        for (int j = 0; j < n - mid; j++) {
            right_arr[j] = arr[j + mid];
        }

        int[] sorted_arr = merge(sort(left_arr), sort(right_arr));

        return sorted_arr;
    }

    public static void printArr(int[] arr) {
        int n = arr.length;
        for (int i = 0; i < n; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        int[] sample_arr = { 32, 23, 20, 17, 13, 10, 7, 5, 1 };
        int[] sorted_arr = sort(sample_arr);
        printArr(sorted_arr);
    }
}

File: Merge sorting algorithm/MergeSort.py
# merge sort is a divide and conquer algorithm
# it divides input  array in two halves 
# calls itself for the two halves
# and then merges the two sorted halves

# two functions are involved in this algorithm
# the Merge() function us used for merging two halves
# the MergeSort() function recusively calles iself to divide the array till size becomes one.
# after dividing the array into smallest units merging starts based on comparison of elements

def merge_two_sorted_lists(a,b):
    a_idx = 0
    b_idx = 0
    sorted_list = []
    
    while (a_idx<len(a) and b_idx<len(b)):
        if(a[a_idx] <= b[b_idx]):
            sorted_list.append(a[a_idx])
            a_idx+=1
        elif(a[a_idx] > b[b_idx]):
            sorted_list.append(b[b_idx])
            b_idx+=1
    
    while(a_idx<len(a)):
        sorted_list.append(a[a_idx])
        a_idx+=1
    
    while(b_idx<len(b)):
        sorted_list.append(b[b_idx])
        b_idx+=1
    
    return sorted_list

def mergesort(arr):
    if(len(arr)<=1):
        return arr

    mid = len(arr) //2

    return merge_two_sorted_lists(mergesort(arr[:mid]),mergesort(arr[mid:]))

if __name__ == "__main__":
    sample_arr = [34,25,21,14,11,10,5,2,1]
    sorted_array = mergesort(sample_arr)
    print(sorted_array)

# to optimal in terms of space complexity you have to reduce the number of arrays used
